method/CellCharter/CellCharter.py

```diff
@@ -0,0 +1,198 @@
+    "--technology",
+    help="The technology of the dataset (Visium, ST, imaging-based).",
+    required=True,
+)
+parser.add_argument(
+    "--seed", help="Seed to use for random operations.", required=True, type=int
+)
+parser.add_argument(
+    "--config",
+    help="Optional config file (json) used to pass additional parameters.",
+    required=False,
+)
+
+################# Get args #######################
+
+args = parser.parse_args()
+
+n_clusters = args.n_clusters
+technology = args.technology
+seed = args.seed
+# Def config
+import json
+with open(args.config, "r") as f:
+    config = json.load(f)
+
+# Set up output files
+from pathlib import Path
+out_dir = Path(args.out_dir)
+label_file = out_dir / "domains.tsv"
+embedding_file = out_dir / "embedding.tsv"
+
+################# Define functions #################
+def get_anndata(args):
+    import anndata as ad
+    import numpy as np
+    import pandas as pd
+    import scipy as sp
+    from PIL import Image
+
+    observations = pd.read_table(args.observations, index_col=0)
+    features = pd.read_table(args.features, index_col=0)
+
+    coordinates = (
+        pd.read_table(args.coordinates, index_col=0)
+        .loc[observations.index, :]
+        .to_numpy()
+    )
+
+    adata = ad.AnnData(obs=observations, 
+                       var=features, 
+                       obsm={"spatial": coordinates})
+
+    if args.matrix is not None:
+        X = sp.io.mmread(args.matrix)
+        if sp.sparse.issparse(X):
+            X = X.tocsr()
+        adata.X = X
+
+    if args.neighbors is not None:
+        adata.obsp["spatial_connectivities"] = sp.io.mmread(args.neighbors).T.tocsr()
+
+    # Filter by selected samples/features
+    if "selected" in adata.obs.columns:
+        adata = adata[observations["selected"].astype(bool), :]
+    if "selected" in adata.var.columns:
+        adata = adata[:, features["selected"].astype(bool)]
+
+    if args.dim_red is not None:
+        adata.obsm["reduced_dimensions"] = (
+            pd.read_table(args.dim_red, index_col=0).loc[adata.obs_names].to_numpy()
+        )
+
+    if args.image is not None:
+        adata.uns["image"] = np.array(Image.open(args.image))
+    else:
+        adata.uns["image"] = None
+
+    return adata
+
+## Your code goes here
+import cellcharter as cc
+import anndata as ad
+import scanpy as sc
+import squidpy as sq
+import pandas as pd
+import numpy as np
+import random
+
+random.seed(seed)
+
+# Load data
+adata = get_anndata(args)
+adata.var_names_make_unique()
+
+# raw count input
+import scvi
+
+# Assume filtered or not?
+# sc.pp.filter_genes(adata, min_counts=3)
+#sc.pp.filter_cells(adata, min_counts=3)
+
+# preprocessing for scvi
+adata.layers["counts"] = adata.X.copy()
+
+# Set up scvi model for reduced_dimension embedding
+scvi.settings.seed = seed
+scvi.model.SCVI.setup_anndata(
+    adata,
+    layer="counts"
+)
+
+model = scvi.model.SCVI(adata)
+model.train(early_stopping=True, enable_progress_bar=True)
```

Comment:
Is this an interactive progress bar? Then it should probably be set to false?

Reply:
Tune them off! :))